modules/Evse15118D20/charger/ISO15118_chargerImpl.cpp

```diff
@@ -44,18 +44,83 @@ iso15118::config::TlsNegotiationStrategy convert_tls_negotiation_strategy(const
 }
 
 types::iso15118_charger::DisplayParameters
-convert_display_parameters(const iso15118::session::feedback::DisplayParameters& in) {
+convert_display_parameters(const iso15118::message_20::DisplayParameters& in) {
+
+    std::optional<float> energy_capacity{std::nullopt};
+    if (in.battery_energy_capacity) {
+        energy_capacity = iso15118::message_20::from_RationalNumber(in.battery_energy_capacity.value());
+    }
+
     return {in.present_soc,
-            in.minimum_soc,
+            in.min_soc,
             in.target_soc,
-            in.maximum_soc,
-            in.remaining_time_to_minimum_soc,
+            in.max_soc,
+            in.remaining_time_to_min_soc,
             in.remaining_time_to_target_soc,
-            in.remaining_time_to_maximum_soc,
-            in.battery_energy_capacity,
+            in.remaining_time_to_max_soc,
+            in.charging_complete,
+            energy_capacity,
             in.inlet_hot};
 }
 
+types::iso15118_charger::DcChargeDynamicModeValues
+convert_dynamic_values(const iso15118::message_20::DC_ChargeLoopRequest::Dynamic_DC_CLReqControlMode& in) {
+
+    using namespace iso15118::message_20;
+
+    std::optional<float> departure_time{std::nullopt};
+    if (in.departure_time) {
+        departure_time = static_cast<float>(*in.departure_time);
+    }
+
+    return {from_RationalNumber(in.target_energy_request),
+            from_RationalNumber(in.max_energy_request),
+            from_RationalNumber(in.min_energy_request),
+            from_RationalNumber(in.max_charge_power),
+            from_RationalNumber(in.min_charge_power),
+            from_RationalNumber(in.max_charge_current),
+            from_RationalNumber(in.max_voltage),
+            from_RationalNumber(in.min_voltage),
+            departure_time,
+            std::nullopt,
+            std::nullopt,
+            std::nullopt,
+            std::nullopt,
+            std::nullopt};
+}
+
+types::iso15118_charger::DcChargeDynamicModeValues
+convert_dynamic_values(const iso15118::message_20::DC_ChargeLoopRequest::BPT_Dynamic_DC_CLReqControlMode& in) {
+
+    using namespace iso15118::message_20;
+
+    std::optional<float> departure_time, max_v2x_req, min_v2x_req{std::nullopt};
+    if (in.departure_time) {
+        departure_time = static_cast<float>(*in.departure_time);
+    }
+    if (in.max_v2x_energy_request) {
+        max_v2x_req = from_RationalNumber(in.max_v2x_energy_request.value());
+    }
+    if (in.min_v2x_energy_request) {
+        min_v2x_req = from_RationalNumber(in.min_v2x_energy_request.value());
+    }
```

Comment:
Just a hint, given that the types have conversion operators defined (which would need to be done for `RationalNumber`) one could define and use such a template:
```
template <typename T, typename U> std::optional<T> convert_from_optional(const std::optional<U>& in) {
  return (in) ? static_cast<T>(*in) : std::nullopt;
}
```
and use it as `departure_time = convert_from_optional<float>(in.departure_time)`

Reply:
Done